sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,308 @@
+	}
+
+	return nil
+}
+
+// estimateSweepFeeIncrement returns fee details for adding the sweep to a batch
+// and for creating new batch with this sweep only.
+func estimateSweepFeeIncrement(s *sweep) (feeDetails, feeDetails, error) {
+	// Create a fake batch with this sweep.
+	batch := &batch{
+		rbfCache: rbfCache{
+			FeeRate: s.minFeeRate,
+		},
+		sweeps: map[lntypes.Hash]sweep{
+			s.swapHash: *s,
+		},
+	}
+
+	// Estimate new batch.
+	fd1, err := estimateBatchWeight(batch)
+	if err != nil {
+		return feeDetails{}, feeDetails{}, err
+	}
+
+	// Add the same sweep again to measure weight increments.
+	swapHash2 := s.swapHash
+	swapHash2[0]++
+	batch.sweeps[swapHash2] = *s
+
+	// Estimate weight of a batch with two sweeps.
+	fd2, err := estimateBatchWeight(batch)
+	if err != nil {
+		return feeDetails{}, feeDetails{}, err
+	}
+
+	// Create feeDetails for sweep.
+	sweepFeeDetails := feeDetails{
+		FeeRate:        s.minFeeRate,
+		NonCoopHint:    s.nonCoopHint,
+		IsExternalAddr: s.isExternalAddr,
+
+		// Calculate sweep weight as a difference.
+		CoopWeight:    fd2.CoopWeight - fd1.CoopWeight,
+		NonCoopWeight: fd2.NonCoopWeight - fd1.NonCoopWeight,
+	}
+
+	return sweepFeeDetails, fd1, nil
+}
+
+// estimateBatchWeight estimates batch weight and returns its fee details.
+func estimateBatchWeight(batch *batch) (feeDetails, error) {
+	// Make sure the batch is not empty.
+	if len(batch.sweeps) == 0 {
+		return feeDetails{}, errors.New("empty batch")
+	}
+
+	// Make sure fee rate is valid.
+	if batch.rbfCache.FeeRate < chainfee.AbsoluteFeePerKwFloor {
+		return feeDetails{}, fmt.Errorf("feeRate is too low: %v",
+			batch.rbfCache.FeeRate)
+	}
+
+	// Find if the batch has at least one non-cooperative sweep.
+	hasNonCoop := false
+	for _, sweep := range batch.sweeps {
+		if sweep.nonCoopHint {
+			hasNonCoop = true
+		}
+	}
+
+	// Find some sweep of the batch. It is used if there is just one sweep.
+	var theSweep sweep
+	for _, sweep := range batch.sweeps {
+		theSweep = sweep
+		break
+	}
+
+	// Find sweep destination address (type) for weight estimations.
+	var destAddr btcutil.Address
+	if theSweep.isExternalAddr {
+		if theSweep.destAddr == nil {
+			return feeDetails{}, errors.New("isExternalAddr=true, " +
+				"but destAddr is nil")
+		}
+		destAddr = theSweep.destAddr
+	} else {
+		// Assume it is taproot by default.
+		destAddr = (*btcutil.AddressTaproot)(nil)
+	}
+
+	// Make two estimators: for coop and non-coop cases.
+	var coopWeight, nonCoopWeight input.TxWeightEstimator
+
+	// Add output weight to the estimator.
+	err := sweeppkg.AddOutputEstimate(&coopWeight, destAddr)
+	if err != nil {
+		return feeDetails{}, fmt.Errorf("sweep.AddOutputEstimate: %w",
+			err)
+	}
+	err = sweeppkg.AddOutputEstimate(&nonCoopWeight, destAddr)
+	if err != nil {
+		return feeDetails{}, fmt.Errorf("sweep.AddOutputEstimate: %w",
+			err)
+	}
+
+	// Add inputs.
+	for _, sweep := range batch.sweeps {
+		// TODO: it should be txscript.SigHashDefault.
```

Comment:
why not sign the whole tx? Maybe leave a comment here for the rationale for `SigHashDefault`.

Reply:
> why not sign the whole tx?

There is no reason to use non-standard sighash here. It was just done this way (SigHashAll) in fee estimation in `publishBatchCoop`, so I did it the same way in the greedy algorithm and left this TODO.

I replaced it later in this PR (in commit "sweepbatcher: fix inaccuracies in fee estimations") and removed the TODO.